### PR TITLE
Store credentials in a cluster safe way

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

Replace the old way of storing credentials for authentication with third party services with a new method that supports clustered couchdb. Adds an API for setting the credentials and adds encryption at rest.

Also solves an issue where the plaintext password could be output in the log files.

medic/cht-core#5904
medic/medic-infrastructure#302

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by garethbowen** [comment]
@m5r Can I get a review please? It's probably easiest to start with the [docs PR](https://github.com/medic/cht-docs/pull/668) to work out what's new.

File: api/src/controllers/credentials.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    const password = req.body;
+
+    if (!key) {
+      return serverUtils.error({ code: 400, reason: 'Missing required param "key"' }, req, res);
+    }
+    if (_.isEmpty(password)) {
+      return serverUtils.error({ code: 400, reason: 'Missing required request body' }, req, res);
+    }
+
+    return checkAuth(req)
```
Comment (m5r) [inline api/src/controllers/credentials.js:29]: I just learned about a database `_security` with this PR and I'm wondering if we need this check if couch already does it for us?

**Review by m5r — CHANGES_REQUESTED** [review]
Great work! I've inlined a few changes requests and one question, nothing major

File: shared-libs/settings/src/index.js [hunk 2]
```diff
@@ -1,39 +1,94 @@
 const request = require('request-promise-native');
-const RESULT_PARSE_REGEX = /^"(.*)"\n?$/;
+const crypto = require('crypto');
 
-// This API gives weird psuedo-JSON results:
-//   "password"\n
-// Should be just `password`
-const parseResponse = response => response.match(RESULT_PARSE_REGEX)[1];
+const IV_LENGTH = 16;
+const CRYTPO_ALGO = 'aes-256-cbc';
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply shared-libs/settings/src/index.js:5]: Good spotting!

File: shared-libs/settings/src/index.js [hunk 3]
```diff
@@ -1,39 +1,94 @@
...
+        // No credentials defined
+        return;
+      }
+      // Throw it regardless so the process gets halted, we just error above for higher specificity
+      throw err;
+    });
+};
+
+const getKey = () => {
+  const url = `${getCouchConfigUrl()}/couch_httpd_auth/secret`;
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply shared-libs/settings/src/index.js:40]: Good point. I've added a comment in the code, and checked that e2e tests will fail if it's not fixed.

**Review by m5r — APPROVED** [review]
LGTM, nice work!

File: shared-libs/settings/src/index.js [hunk 4]
```diff
@@ -1,39 +1,96 @@
...
+  // https://docs.couchdb.org/en/stable/config/auth.html#chttpd_auth/secret
+  const url = `${getCouchConfigUrl()}/couch_httpd_auth/secret`;
+  return request.get(url, { json: true });
+};
+
+const encrypt = (text) => {
+  return getKey()
+    .then(key => {
+      const iv = crypto.randomBytes(IV_LENGTH);
+      const cipher = crypto.createCipheriv(CRYPTO_ALGO, Buffer.from(key), iv);
```
Comment (dianabarsan) [inline shared-libs/settings/src/index.js:50]: I'm trying this locally, with Node 16, and I'm getting this error: 
```
2022-06-10 08:05:09 [33mWARN[39m: Non-numeric error code: ERR_CRYPTO_INVALID_KEYLEN 
2022-06-10 08:05:09 [31mERROR[39m: Server error: RangeError: Invalid key length
    at Cipheriv.createCipherBase (node:internal/crypto/cipher:116:19)
    at Cipheriv.createCipherWithIV (node:internal/crypto/cipher:135:3)
    at new Cipheriv (node:internal/crypto/cipher:243:3)
    at Object.createCipheriv (node:crypto:138:10)
    at /shared-libs/settings/src/index.js:49:29
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 1) {
  [stack]: 'RangeError: Invalid key length\n' +
    '    at Cipheriv.createCipherBase (node:internal/crypto/cipher:116:19)\n' +
    '    at Cipheriv.createCipherWithIV (node:internal/crypto/cipher:135:3)\n' +
    '    at new Cipheriv (node:internal/crypto/cipher:243:3)\n' +
    '    at Object.createCipheriv (node:crypto:138:10)\n' +
    '    at /shared-libs/settings/src/index.js:49:29\n' +
    '    at runMicrotasks (<anonymous>)\n' +
    '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
    '    at async Promise.all (index 1)',
  [message]: 'Invalid key length',
  code: 'ERR_CRYPTO_INVALID_KEYLEN'
} 
```

My couchdb secret is `6c1953b6-e64d-4b0c-9268-2528396f2f58`. 

The internet says that the buffer length is the problem. I hacked by by doing: 
```
const cipher = crypto.createCipheriv(CRYPTO_ALGO, Buffer.from(key).slice(0, 32), iv);
const decipher = crypto.createDecipheriv(CRYPTO_ALGO, Buffer.from(key).slice(0, 32), iv);
```

Comment (garethbowen) [reply shared-libs/settings/src/index.js]: @dianabarsan Great catch, thanks! Fixed and integration test added.

**Comment by dianabarsan** [comment]
I'm needing this for ArchV3. Could we AT this and merge this to master, so I have a clean commit to cherry pick?

**Comment by garethbowen** [comment]
@dianabarsan Sorry, I didn't realise it was blocking! Marked for AT.

**Comment by dianabarsan** [comment]
No apology needed :)

**Pull request merged**